modules/services/window-managers/hyprland.nix

```
@@ -0,0 +1,129 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.wayland.windowManager.hyprland;
+  defaultHyprlandPackage = pkgs.hyprland.override {
+    enableXWayland = cfg.xwayland.enable;
+    hidpiXWayland = cfg.xwayland.hidpi;
+    nvidiaPatches = cfg.nvidiaPatches;
+  };
+in {
+  meta.maintainers = [ lib.maintainers.fufexan ];
+
+  options.wayland.windowManager.hyprland = {
+    enable = lib.mkEnableOption (lib.mdDoc "Hyprland wayland compositor");
+
+    package = lib.mkOption {
+      type = with lib.types; nullOr package;
+      default = defaultHyprlandPackage;
+      description = lib.mdDoc ''
+        Hyprland package to use. Will override the 'xwayland' and 'nvidiaPatches'
+        options.
+      '';
+    };
+
+    plugins = lib.mkOption {
+      type = with lib.types; listOf (either package path);
+      default = [ ];
+      description = lib.mdDoc ''
+        List of Hyprland plugins to use. Can either be packages or
+        absolute plugin paths.
+      '';
+    };
+
+    systemdIntegration = lib.mkOption {
+      type = lib.types.bool;
+      default = pkgs.stdenv.isLinux;
+      description = lib.mdDoc ''
+        Whether to enable {path}`hyprland-session.target` on
+        hyprland startup. This links to `graphical-session.target`.
+        Some important environment variables will be imported to systemd
+        and dbus user environment before reaching the target, including
+        - `DISPLAY`
+        - `HYPRLAND_INSTANCE_SIGNATURE`
+        - `WAYLAND_DISPLAY`
+        - `XDG_CURRENT_DESKTOP`
+      '';
+    };
+
+    disableAutoreload = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      example = lib.literalExpression "true";
+      description = lib.mdDoc ''
+        Whether to disable automatically reloading Hyprland's configuration when
+        rebuilding the Home Manager profile.
+      '';
+    };
+
+    xwayland = {
+      enable = lib.mkEnableOption (lib.mdDoc "XWayland") // { default = true; };
+      hidpi = lib.mkEnableOption null // {
+        description = lib.mdDoc ''
+          Enable HiDPI XWayland, based on [XWayland MR 733](https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/733).
+          See <https://wiki.hyprland.org/Nix/Options-Overrides/#xwayland-hidpi> for more info.
+        '';
+      };
+    };
+
+    nvidiaPatches = lib.mkEnableOption
+      (lib.mdDoc "patching wlroots for better Nvidia support.");
+
+    extraConfig = lib.mkOption {
+      type = lib.types.nullOr lib.types.lines;
```

I think you could omit the `null` case.

```suggestion
      type = lib.types.lines;
```

That said, to me it seems like it would be possible to relatively easily create an RFC-42 style `settings` option instead of `extraConfig`?

---

I think it could work. Currently there's https://github.com/hyprwm/Hyprland/pull/870 which is *very* big and intricate.

Perhaps adapting `lib.generators.toINI` could work, but I'm unsure how to actually do it. I've tried 
```nix
lib: let
  libStr = lib.strings;
  libAttr = lib.attrsets;
in {
  toHyprconf = {
    # format a setting line from key and value
    mkKeyValue ? lib.generators.mkKeyValueDefault {} "=",
    # allow lists as values for duplicate keys
    listsAsDuplicateKeys ? false,
  }: attrsOfAttrs: let
    # map function to string for each key val
    mapAttrsToStringsSep = sep: mapFn: attrs:
      libStr.concatStringsSep sep
      (libAttr.mapAttrsToList mapFn attrs);
    mkSection = sectName: sectValues:
      ''
        ${sectName} {
      ''
      + (lib.generators.toKeyValue {inherit mkKeyValue listsAsDuplicateKeys;} sectValues)
      + ''
        }
      '';
  in
    # map input to hyprconf sections
    mapAttrsToStringsSep "\n" mkSection attrsOfAttrs;
}
```
but couldn't test it in a repl.

---

Here is a rough and mostly untested attempt:

```nix
toHyprconf = { }: attrs: let
  mkSection = n: attrs: ''
    ${n} {
    ${toKeyValue { listsAsDuplicateKeys = true; } attrs}
    }
  '';

  mkField = mkKeyValueDefault {} "=";

  sections = filterAttrs (n: v: isAttrs v) attrs;
  fields = filterAttrs (n: v: !(isAttrs v)) attrs;
in
  concatStringsSep "\n" (
    mapAttrsToList mkSection sections ++ mapAttrsToList mkField fields
  );
```

This does assume that Hyprland does not care about field order, but I think that is a correct assumption?

---

That gets us closer, but there are instances of nested sections, for example
```ini
input {
  sensitivity = 0.5
  touchpad {
    natural_scroll = true
  }
}

---

Maybe something like

```nix
toHyprconf = { }: attrs: let
  mkSection = n: attrs: ''
    ${n} {
    ${toHyprconf { } attrs}}
  '';

  mkFields = generators.toKeyValue { listsAsDuplicateKeys = true; };

  sections = filterAttrs (n: v: isAttrs v) attrs;
  fields = filterAttrs (n: v: !(isAttrs v)) attrs;
in
  concatStringsSep "\n" (mapAttrsToList mkSection sections)
  + mkFields fields;
```

Won't do indentation or anything fancy but I guess good enough to start with.

---

That looks promising. I'll play around to see whether I can get it to indent properly. Thanks!

---

The hyprland config allows a key to have multiple values. For example
```ini
exec-once=eww open bar
exec-once=firefox

bind=SUPER, Return, exec, kitty
bind=SUPER, Return, exec, notify-send "Kitty was started!"
```

How could we make this work?

---

That's what `listsAsDuplicateKeys` does

---

That's really handy then!

---

Sorry for the long delay, I've been having exams for the past month.

I just realized we have to handle submaps (nested keybinds) as a special case.
```ini
    # window resize
    bind = $mod, S, submap, resize

    submap = resize
    binde = , right, resizeactive, 10 0
    binde = , left, resizeactive, -10 0
    binde = , up, resizeactive, 0 -10
    binde = , down, resizeactive, 0 10
    bind = , escape, submap, reset
    submap = reset
```

Here, the order of binds matters. Any binds between `submap = ...`. Can we simply pass this as `key=value` (I'm thinking multiline string) or is there a better way?

---

It seems the multiline string way worked just fine, although it looks a bit ugly.